File: alds1/9b_maximum_heap.c
#define dbg(fmt,...) fprintf(stderr,fmt,__VA_ARGS__)
#define dpri(x) dbg(#x ": %d\n", x)
#define dprs(x) dbg(#x ": %s\n", x)
#include <stdio.h>
typedef long long ll;
const int MYINF = 1e9+7;
typedef int bool;
const bool true = 1;
const bool false = 0;

#define MAX_H 500010
#define parent(i) ((i)/2)
#define left(i) ((i)*2)
#define right(i) ((i)*2+1)

int H;
int heap[MAX_H];
#define len heap[0]

void maxHeapify(int i) {
	int l, r, *A, largest;
	A = heap;
	l = left(i);
	r = right(i);

	if (l <= H && A[l] > A[i]) largest = l;
	else largest = i;

	if (r <= H && A[r] > A[largest]) largest = r;

	if (largest != i) {
		int t;
		t = A[i];
		A[i] = A[largest];
		A[largest] = t;
		maxHeapify(largest);
	}
}

void buildMaxHeap() {
	int i;
	for (i = H/2; 0 < i; i--) {
		maxHeapify(i);
	}
}

void insert(int x) {
	heap[++len] = x;
}

//void insert(int x) {
//	int i, t;
//	heap[++len] = x;
//	for (i = len; 0 < parent(i) && heap[parent(i)] < heap[i]; i = parent(i)) {
//		t = heap[parent(i)];
//		heap[parent(i)] = heap[i];
//		heap[i] = t;
//	}
//}

int main(int argc, char **argv)
{
	int i, x;
	scanf("%d", &H);
	for (i = 0; i < H; i++) {
		scanf("%d", &x);
		insert(x);
	}
	buildMaxHeap();

	for (i = 1; i <= H; i++) {
		printf(" %d", heap[i]);
	}
	puts("");

	return 0;
}

File: alds1/9b_maximum_heap.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"unicode"
)

func parent(i int) int {
	return i / 2
}

func left(i int) int {
	return i * 2
}

func right(i int) int {
	return i*2 + 1
}

func maxHeapify(a []int, i int) {
	l := left(i)
	r := right(i)
	h := len(a) - 1
	largest := -1

	if l <= h && a[l] > a[i] {
		largest = l
	} else {
		largest = i
	}
	if r <= h && a[r] > a[largest] {
		largest = r
	}

	if largest != i {
		a[i], a[largest] = a[largest], a[i]
		maxHeapify(a, largest)
	}
}

func main() {
	defer stdout.Flush()
	n := readInt()
	a := make([]int, n+1)
	for i := 1; i <= n; i++ {
		a[i] = readInt()
	}
	for i := n / 2; i > 0; i-- {
		maxHeapify(a, i)
	}
	for i := 1; i <= n; i++ {
		printf(" %d", a[i])
	}
	println()
}

var (
	stdin  *bufio.Reader
	stdout *bufio.Writer
)

func init() {
	stdin = bufio.NewReader(os.Stdin)
	stdout = bufio.NewWriter(os.Stdout)
}

func scanf(f string, args ...interface{}) (int, error) {
	n, err := fmt.Fscanf(stdin, f, args...)
	for {
		r, _, err := stdin.ReadRune()
		if err != nil || !unicode.IsSpace(r) {
			stdin.UnreadRune()
			break
		}
	}
	return n, err
}

func readInt() int {
	var n int
	scanf("%d", &n)
	return n
}

func readIntSlice(n int) []int {
	a := make([]int, n)
	for i := 0; i < n; i++ {
		scanf("%d", &a[i])
	}
	return a
}

func readLengthAndSlice() []int {
	n := readInt()
	return readIntSlice(n)
}

func readString() string {
	var s string
	scanf("%s", &s)
	return s
}

func printf(f string, args ...interface{}) (int, error) {
	return fmt.Fprintf(stdout, f, args...)
}

func println(args ...interface{}) (int, error) {
	return fmt.Fprintln(stdout, args...)
}
